timemachine/fe/dummy.py

```diff
@@ -0,0 +1,618 @@
+    for i in range(N):
+        if i not in core:
+            induced_g.add_node(i)
+            for j in range(i + 1, N):
+                if j not in core:
+                    dist = nx.shortest_path_length(g, source=i, target=j)
+                    if dist <= 2:
+                        induced_g.add_edge(i, j)
+
+    cc = list(nx.connected_components(induced_g))
+
+    # check that we didn't miss any dummy atoms
+    assert np.sum([len(c) for c in cc]) == (N - len(core))
+
+    return cc
+
+
+def enumerate_dummy_ixns(dg, ag):
+    """
+    Enumerate the allowed set of interactions between
+    1) atoms within the dummy group
+    2) atoms in a dummy group and atoms in anchor group
+
+    Parameters
+    ----------
+    dg: set of int
+        dummy atoms in a dummy group
+
+    ag: list of int
+        ordered core anchor atoms
+
+    Returns
+    -------
+    set
+        Set of interactions containing, 1-2, 1-3, and 1-4 interactions
+        The bonds returned are ordered.
+
+    """
+
+    allowed_ixns = set()
+
+    nc1 = list(itertools.combinations(dg, 1))
+    nc2 = list(itertools.combinations(dg, 2))
+    nc3 = list(itertools.combinations(dg, 3))
+    nc4 = list(itertools.combinations(dg, 4))
+
+    for ijkl in nc4:
+        for i, j, k, l in itertools.permutations(ijkl):
+            allowed_ixns.add((i, j, k, l))
+    for ijk in nc3:
+        for i, j, k in itertools.permutations(ijk):
+            allowed_ixns.add((i, j, k))
+    for ij in nc2:
+        for i, j in itertools.permutations(ij):
+            allowed_ixns.add((i, j))
+
+    # TBD: restrict to terms where bonds/angles actually exist
+
+    if len(ag) > 0:
+        a = ag[0]
+        for ijk in nc3:
+            for i, j, k in itertools.permutations(ijk):
+                allowed_ixns.add((a, i, j, k))
+                allowed_ixns.add((i, a, j, k))
+                allowed_ixns.add((i, j, a, k))
+                allowed_ixns.add((i, j, k, a))
+        for ij in nc2:
+            for i, j in itertools.permutations(ij):
+                allowed_ixns.add((a, i, j))
+                allowed_ixns.add((i, a, j))
+                allowed_ixns.add((i, j, a))
+
+        for (i,) in nc1:
+            allowed_ixns.add((i, a))
+            allowed_ixns.add((a, i))
+            # this is actually same as above
+            # included to please the symmetry gods
+
+        if len(ag) > 1:
+            b = ag[1]
+            for ij in nc2:
+                for i, j in itertools.permutations(ij):
+                    allowed_ixns.add((i, j, a, b))
+                    # this looks weird at first, but is actually valid
+                    # consider:
+                    # i j
+                    # |/
+                    # a-b
+                    # we can freely rotate i-j independently of a-b
+                    # in both of the following cases.
+                    allowed_ixns.add((i, a, b, j))
+                    allowed_ixns.add((a, b, i, j))
```

Comment:
lmk if this is still not clear - trying to resolve all outstanding issues

Reply:
> To express this in code rather than comments, we might need to add more geometric / symmetry-flavored types than tuples

Just wanted to +1 this idea (probably for future PR) -- seems like it could help with readability to give names to the different interaction tuples and encode the equivalence logic close to the type definition

Reply:
This has been simplified a bit since we no longer allow non-strict stuff. The important thing to remember is 1-2, 1-3, 1-4 take in ordered indices but I should be clearer about that in the comments.